src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -528,3 +570,77 @@ object IndexLogEntry {
 
   def schemaString(schema: StructType): String = schema.json
 }
+
+/**
+ * Provides functionality to generate unique file ids for files.
+ */
+class FileIdTracker {
+  private var maxId: Long = -1L
+
+  // Combination of file properties, used as key, to identify a
+  // unique file for which an id is generated.
+  type key = (
+    String, // Full path.
+      Long, // Size.
+      Long  // Modified time.
+    )
+  private val fileToIdMap: mutable.HashMap[key, Long] = mutable.HashMap()
+
+  def getMaxFileId: Long = maxId
+
+  def getFileToIdMap: HashMap[key, Long] = fileToIdMap
+
+  def getFileId(path: String, size: Long, modifiedTime: Long): Option[Long] =
+    fileToIdMap.get((path, size, modifiedTime))
+
+  def setSizeHint(size: Int): Unit = fileToIdMap.sizeHint(size)
+
+  /**
+   * Add a set of FileInfos to the fileToIdMap. The assumption is
+   * that the each FileInfo already has a valid file id if an entry
+   * with that key already exists in the fileToIdMap, then it has
+   * the same file id (i.e. no new file id is generated and only
+   * maxId is updated according to the new entries).
+   *
+   * @param files Set of FileInfo instances to add to the fileToIdMap.
+   */
+  def addFileInfo(files: Set[FileInfo]): Unit = {
+    setSizeHint(files.size)
+    files.foreach { f =>
+      val key = (f.name, f.size, f.modifiedTime)
+      fileToIdMap.get(key) match {
+        case Some(id) =>
+          if (id != f.id) {
+            throw HyperspaceException(
+              "Adding file info with a conflicting id. " +
+                s"(existing id: $id, new id: ${f.id}, file: ${f.name}).")
+          }
+        case _ =>
+          if (f.id == IndexConstants.UNKNOWN_FILE_ID) {
+            throw HyperspaceException(
+              s"Can not add file info with unknown id. (file: ${f.name}).")
+          }
+
+          fileToIdMap.put(key, f.id)
+          maxId = Math.max(maxId, f.id)
+      }
+    }
+  }
```

Comment:
How about the following? Note that `put` returns `None` if no matching key exists.
```suggestion
  def addFileInfo(files: Set[FileInfo]): Unit = {
    setSizeHint(files.size)
    files.foreach { f =>
      if (f.id == IndexConstants.UNKNOWN_FILE_ID) {
        throw HyperspaceException(
          s"Cannot add file info with unknown id. (file: ${f.name}).")
      }

      val key = (f.name, f.size, f.modifiedTime)
      fileToIdMap.put(key, f.id) match {
        case Some(id) =>
          if (id != f.id) {
            throw HyperspaceException(
              "Adding file info with a conflicting id. " +
                s"(existing id: $id, new id: ${f.id}, file: ${f.name}).")
          }
        case None =>
          maxId = Math.max(maxId, f.id)
      }
    }
  }
```